pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java

```diff
@@ -84,8 +92,19 @@ public boolean send(TransferableBlock block)
     if (block.isSuccessfulEndOfStreamBlock()) {
       // Send metadata to only one randomly picked mailbox, and empty EOS block to other mailboxes
       int numMailboxes = _sendingMailboxes.size();
-      int mailboxIdToSendMetadata = ThreadLocalRandom.current().nextInt(numMailboxes);
-      assert block.getQueryStats() != null;
+      int mailboxIdToSendMetadata;
+      if (block.getQueryStats() != null) {
+        mailboxIdToSendMetadata = ThreadLocalRandom.current().nextInt(numMailboxes);
+        if (LOGGER.isTraceEnabled()) {
+          LOGGER.trace("Sending EOS metadata. Only mailbox #{} will get stats", mailboxIdToSendMetadata);
+        }
+      } else {
+        if (LOGGER.isTraceEnabled()) {
+          LOGGER.trace("Sending EOS metadata. No stat will be sent");
+        }
+        // this may happen when the block exchange is itself used as a sending mailbox, like when using spools
+        mailboxIdToSendMetadata = -1;
```

Comment:
I'm not sure I followed this part - shouldn't the block exchange sending mailbox send the query stats to all the mailboxes in the next layer where the actual random mailbox picking can happen (as an aside, I didn't quite get the purpose of this random mailbox picking logic for query stats either)?

Reply:
I think it is easier to see with a diagram. This is how data is distributed:

```mermaid
flowchart BT
  SendOperator --> BlockExchange
  BlockExchange -- broadcast --> ActualSender1
  BlockExchange -- broadcast --> ActualSender2
  ActualSender1 -- distributionSpecific --> ReceiveOperator1.w1
  ActualSender1 -- distributionSpecific --> ReceiveOperator1.w2
  ActualSender2 -- distributionSpecific --> ReceiveOperator2.w1
  ActualSender2 -- distributionSpecific --> ReceiveOperator2.w2
```

This is how stats are distributed:

```mermaid
flowchart BT
  SendOperator --> BlockExchange
  BlockExchange -- random (single) --> ActualSender1
  BlockExchange -- random (single) --> ActualSender2
  ActualSender1 -- random (single) --> ReceiveOperator1.w1
  ActualSender1 -- random (single) --> ReceiveOperator1.w2
  ActualSender2 -- random (single) --> ReceiveOperator2.w1
  ActualSender2 -- random (single) --> ReceiveOperator2.w2
```

So only 1 worker will receive the stats. For example:

```mermaid
flowchart BT
  SendOperator --> BlockExchange
  BlockExchange -- no stats --> ActualSender1
  BlockExchange -- stats --> ActualSender2
  ActualSender1 -- no stats --> ReceiveOperator1.w1
  ActualSender1 -- no stats --> ReceiveOperator1.w2
  ActualSender2 -- stats --> ReceiveOperator2.w1
  ActualSender2 -- no stats --> ReceiveOperator2.w2

  linkStyle 0 stroke-width:2px,fill:none,stroke:red;
  linkStyle 2 stroke-width:2px,fill:none,stroke:red;
  linkStyle 5 stroke-width:2px,fill:none,stroke:red;
```

This is required because stats are later aggregated to be presented to the user as a tree where worker information is aggregated (usually added).

Reply:
Wow thanks, this makes things very clear 😄 

Would be great to include this in the dev docs if / when we have that!